### ISSUE TITLE
Fix/Train

### ISSUE THREAD

**Issue by TommyBsk** [TommyBsk/Tiny_model_4_CD#2]
Ciao Andrea, 
Dovrei aver sistemato tutta la parte di creazione di dataset e dataloader, infatti ora il train inizia correttamente.
Il problema però è che la rete considera immagini a 3 canali mentre ora le immagini sono grayscale, qunidi sto cercando di sistemare le varie dimensioni dei canali in ingresso.

Buona serata.

**Comment by AndreaCodegoni** [comment]
Se l'immagine che carichi ha come shape (H,W) puoi fare così:

`
b = img_grayscale.unsqueeze(0).repeat(3, 1, 1)
`
se invece ha la shape (1,H,W)

`
b = img_grayscale.repeat(3, 1, 1)
`

Altra alternativa aggiungi un layer iniziale tipo convoluzione 1x1 che passa da 1 canale a 3. 

Buona serata

**Comment by TommyBsk** [comment]
Perfetto risolto.
Grazie mille.

**Comment by AndreaCodegoni** [comment]
Pusha e fai un training di prova e vediamo chesuccede

**Comment by TommyBsk** [comment]
Ho provato a fare un training di prova sul mio computer utilizzando le slice di un solo paziente e questi sono i risultati che ho ottenuto:

Epoch 0
Training phase summary
Loss for epoch 0 is 0.17735304793014245
IoU class change for epoch 0 is 0.0
F1 class change for epoch 0 is 0.0

Validation phase summary
Loss for epoch 0 is 0.015035600091020266
IoU class change for epoch 0 is 0.0
F1 class change for epoch 0 is 0.0

Epoch 1
Training phase summary
Loss for epoch 1 is 0.011716141009374577
IoU class change for epoch 1 is 0.0
F1 class change for epoch 1 is 0.0

Validation phase summary
Loss for epoch 1 is 0.008841802417818043
IoU class change for epoch 1 is 0.0
F1 class change for epoch 1 is 0.0

Epoch 2
Training phase summary
Loss for epoch 2 is 0.00799211732331006
IoU class change for epoch 2 is 0.0
F1 class change for epoch 2 is 0.0

Validation phase summary
Loss for epoch 2 is 0.005414617092659076
IoU class change for epoch 2 is 0.0
F1 class change for epoch 2 is 0.0

Epoch 3
Training phase summary
Loss for epoch 3 is 0.005838834523113773
IoU class change for epoch 3 is 0.003073386472092992
F1 class change for epoch 3 is 0.006127938677984714

Validation phase summary
Loss for epoch 3 is 0.004678304307162762
IoU class change for epoch 3 is 0.021107686559959257
F1 class change for epoch 3 is 0.04134271903607217

Epoch 4
Training phase summary
Loss for epoch 4 is 0.00470171553110156
IoU class change for epoch 4 is 0.18619251818279997
F1 class change for epoch 4 is 0.313933012812488

Validation phase summary
Loss for epoch 4 is 0.004238289134162996
IoU class change for epoch 4 is 0.4676270136264138
F1 class change for epoch 4 is 0.6372558780705322

Epoch 5
Training phase summary
Loss for epoch 5 is 0.0037444032393121984
IoU class change for epoch 5 is 0.5086232850186952
F1 class change for epoch 5 is 0.6742879365572523

Validation phase summary
Loss for epoch 5 is 0.0030392008807717096
IoU class change for epoch 5 is 0.6448658724307582
F1 class change for epoch 5 is 0.7840953288208521

Epoch 6
Training phase summary
Loss for epoch 6 is 0.0037009192815543536
IoU class change for epoch 6 is 0.536747427273641
F1 class change for epoch 6 is 0.6985499010269098

Validation phase summary
Loss for epoch 6 is 0.0036961951474141744
IoU class change for epoch 6 is 0.43918065330695233
F1 class change for epoch 6 is 0.6103203436476403

Epoch 7
Training phase summary
Loss for epoch 7 is 0.0036715559510733277
IoU class change for epoch 7 is 0.5313952992350948
F1 class change for epoch 7 is 0.694001417481193

Validation phase summary
Loss for epoch 7 is 0.003371319060938226
IoU class change for epoch 7 is 0.5768212842966456
F1 class change for epoch 7 is 0.7316253822585552

**Comment by AndreaCodegoni** [comment]
Bene, ora che funziona trasferisci sul cluster e prova a fare qualche esperimento. Hai un po' di parametri su cui giocare come le augmentations, i parametri dell'ottimizzatore e il modello. Se riesci cerca di fare di salvarti anche delle immagini carine ogni tanto così vediamo cosa effettivamente sputa fuori.

**Comment by TommyBsk** [comment]
Ok perfetto, provo così.

**Comment by TommyBsk** [comment]
Ciao Andrea, 

Ho provato a fare andare la rete sul cluster ma mi sono accorto di un problema sulla dimensione delle immagini.
I volumi che ci arrivano hanno dimensioni differenti nei 3 assi, ad esempio uno dei volumi ha dimensione 512x512x625, altre 512x512x400.
Con Gabriele avevamo fatto un downsample di tutto il volume riducendolo di 2 con l'obiettivo di portare le immagini ad avere una dimensione di 256x256.
Questo funziona correttamente per tutte le immagini sull'asse X mentre per le immagini su asse Y e Z ci sono dei problemi perchè le immagini hanno dimensione 313x256 oppure 200x256 e questo non permette di metterle in uno stack per effettuare il training della rete.
Ho letto che il problema è nello stack del dataloader e quindi una possibile soluzione potrebbe essere creare uno batch nel getitem del dataset e usare una batch_size = 1 nel dataloader.
Oppure cercare di fare il downsample sulle immagini bidimensionali invece che su volumi in modo da poter trasformarle tutte in 256x256.

**Comment by AndreaCodegoni** [comment]
Ciao Tommaso,

la strada di utilizzare reti differenti per i singoli assi la possiamo sperimentare ma non la terrei come prima scelta.
Io inizierei con il provare un training solo per l'asse X che per il momento è già pronto in modo da iniziare a vedere cosa otteniamo.

Mentre vanno i training puoi provare la soluzione semplice ed aggiungere una trasformazione shared nelle augmentation che è quella della Resize di albumentations 
```
def _create_shared_augmentation():
    return alb.Compose(
        [
            alb.Resize(256,256)
            alb.Flip(p=0.5),
            alb.Rotate(limit=5, p=0.5),
        ],
        additional_targets={"image0": "image", "x_mask0": "mask"},
    )
```
Questa dovrebbe mettere una pezza anche se magari non è il massimo. Però visto che si tratta di una prova minima la farei e proverei a fare qualche training per vedere i risultati. 

Poi si sono d'accordo, riguardiamo insieme la prossima volta la data prepare e vediamo quali passaggi fare nella prima suddivisione dei volumi e quali fare nel dataset. Una possibilità è quella di salvarsi per tutti gli assi le slice facendo solo le normalizzazioni tra 0-1, e poi le questioni di clahe e resize farle nel dataset in modo da poter decidere poi la risoluzione migliore.

I training falli almeno per un 20/30 epoche. Così capiamo se stiamo overfittando.

**Comment by TommyBsk** [comment]
Ho fatto un train con 40 epoche su tutte le assi aggiungendo il resize e questi sono i risultati:
[slurm-29713.txt](https://github.com/TommyBsk/Tiny_model_4_CD/files/11991352/slurm-29713.txt)

**Comment by AndreaCodegoni** [comment]
Bene mi sembra che inizino ad essere decenti come risultati.
Come prima cosa ti direi di iniziare a fare uno script per salvarti delle immagini da far vedere (non 10 per singola immagine, massimo 2/3, puoi giocare con la batchsize nello script di test). In questa settimana cerco di fissare un incontro in cui parteciperà anche Gabriele in modo che lo aggiorniamo e gli facciamo vedere un po' di immagini e risultati.

Fino ad allora ti direi cerca di prendere bene confidenza con il codice così che poi magari proviamo a fare modifiche per fare sempre meglio mentre iniziamo anche a pensare a cosa fare per gli step successivi.

Buona domenica

**Comment by TommyBsk** [comment]
Ok perfetto, quindi 2/3 immagini per training?
Volendo potrei scegliere delle immagini un po' più significative dal set di evaluetion oppure dal paziente che non viene utilizzato.
Buona domenica.

**Comment by AndreaCodegoni** [comment]
Si il meglio sarebbe avere CT/GT/previsione_del_modello sul validation o sul test set e magari lo score proprio sul test set mai visto. 

2/3 era inteso come organizzare le terne CT/GT/previsione_del_modello in una singola immagine che ti salvi per poi vederle, perché se ogni immagine che ti salvi ne contiene di più non apprezziamo molto essendo cose piccole. Quante salvarne in totale vedi tu. Puoi salvarle anche tutte e poi prendiamo quelle più significative da far vedere, anche quando saranno da mettere nella tesi.

**Comment by TommyBsk** [comment]
Ah ok perfetto, non avevo capito.
Allora provo così.
Grazie mille.

**Comment by AndreaCodegoni** [comment]
Ottimo dai! 

A presto.

**Comment by TommyBsk** [comment]
Ciao Andrea, 
Il cluster oggi è tornato a funzionare e ho fatto qualche train per produrre delle immagini, in realtà le immagini che escono dai dati di test non sono bellissime anche se le metriche sono abbastanza alte, invece le immagini prodotte su dati usati per il training vengono meglio.
![prediction_image_1](https://github.com/TommyBsk/Tiny_model_4_CD/assets/97763318/0d5ec6f4-8a3d-4cf3-b5b4-1a2e62594b00)
![prediction_image_3](https://github.com/TommyBsk/Tiny_model_4_CD/assets/97763318/dda3ee59-b6f0-459f-842e-b9108753cde7)
![prediction_image_2](https://github.com/TommyBsk/Tiny_model_4_CD/assets/97763318/bece59c6-d81c-4464-a481-b1b0f4a7f1d6)
[metrics.txt](https://github.com/TommyBsk/Tiny_model_4_CD/files/12028330/metrics.txt)

Queste sono le metriche sul test, ho usato solo 9 immagini per provare se tutto funzionasse
'acc': 0.9973070780436175, 'miou': 0.6439573856546331, 'mf1': 0.7244980558966503, 'iou_0': 0.9973041039556699, 'iou_1': 0.29061066735359625, 'F1_0': 0.9986501729569129, 'F1_1': 0.4503459388363875, 'precision_0': 0.9975727910947362, 'precision_1': 0.8039375017806313, 'recall_0': 0.999730003954037, 'recall_1': 0.31277836014998983}

Ho provato anche test_ondata e funziona tutto correttamente anche la parte di ricostruzione del volume.
Buona giornata.

**Comment by AndreaCodegoni** [comment]
Ciao Tommaso,

dalle immagini che hai riportato noto che ci sono ancora delle cattive corrispondenze tra le CT e le maschere. La prima e la terza sono abbastanza evidenti. Questo vuol dire che il modello in training sta overfittando tantissimo per riprodurti maschere ma non sta trovando nulla sta solo riproducendo le maschere che vede.  Le metriche sul test infatti sono basse, devi guardare quelle con "_1", tipo  'F1_1': 0.4503459388363875

Devi ricontrollare ancora i dati. Per farlo fai uno script che ti carica la CT e la maschera, e poi tu produce dei plot di CT originale, maschera e poi CT contro maschera per vedere se appunto la maschera corrisponde bene alla CT.

**Comment by TommyBsk** [comment]
In realtà, guardandole così sembravano anche a me sbagliate le corrispondenze e ho controllato e dovrebbero essere corrette.
Una cosa che non mi sembra corretto è l'allineamente tra gt e slice ma ho provato a sovrapporle con un editor di immagini e dovrebbe essere corretto.
Alcune delle segmentazioni non sono bellissime perchè avevo fatto delle sgementazioni a 3 classi inizialmente e non avevo segmentato le pareti della protesi, quindi poi rendendo la segmentazione binaria ci sono alcuni buchi.

**Comment by AndreaCodegoni** [comment]
Quelle che mi destano perplessità sono ad esempio queste due
<img width="845" alt="Screenshot 2023-07-12 155739" src="https://github.com/TommyBsk/Tiny_model_4_CD/assets/54067932/29c73a9e-cd2a-439b-a8ae-a4f48c7d2414">
<img width="452" alt="Screenshot 2023-07-12 160041" src="https://github.com/TommyBsk/Tiny_model_4_CD/assets/54067932/1b34a4e1-92de-4e26-b1c2-cd342110f6ff">

Mi confermi che sto sbagliando io a fare queste grossolane corrispondenze tirando la linea rossa dove mi sembra che ci siano le cose invertite? mi verrebbe da dire che ci sono due inversioni, sia orizzontale che verticale.

**Comment by TommyBsk** [comment]
Viste così direi di si, ora provo a tirarne giù un po' e provare a guardarne il più possibile per ceracre di capire.

**Comment by AndreaCodegoni** [comment]
Ti consiglio di fare la sovrapposizione come ti dicevo prima tramite script in modo che le cose siano riproducibili e  non si hanno errori di sovrapposizione manuale ed in modo che si può provare a ruotare e capire cosa si sta sbagliando. Possibile anche che l'errore sia nel dataset quando carichi e fai le augmentations. C'è un po' da indagare.

**Comment by TommyBsk** [comment]
Okok allora farò uno scirpt.
Ho provate a guardarne alcune e mi sono accorto che in effetti il problema è solo sull'asse x.
Le machere escono ruotate dalla data prepare.
Sto cercando di capire il motivo.
